### PR TITLE
Add Python version of fpgadiag

### PR DESCRIPTION
**Pull request by r-rojo**
This adds the Python version of fpgadiag. The main files that implement fpgadiag are in the opae.tools.fpgadiag namespace
* fpgadiag.py - entry point
* diagtest.py - basic diagtest (contains logic for nlb0)
* nlb0.py - derives from diagtest, defines its guid
* nlb3.py - derives from diagtest, defines its guid, and overloads some functions
* nlb7.py - derives from diagtest, defines its guid, and overloads some functions

This also adds some utility classes in the opae.utils namespace

### CODE REVIEW

File: tools/extra/pydiag/opae/tools/fpgadiag/nlb0.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Copyright(c) 2017, Intel Corporation
```
Comment (tswhison) [inline tools/extra/pydiag/opae/tools/fpgadiag/nlb0.py]: s/2017/2018/

File: tools/extra/pydiag/opae/tools/fpgadiag/nlb3.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
+# Copyright(c) 2017, Intel Corporation
```
Comment (tswhison) [inline tools/extra/pydiag/opae/tools/fpgadiag/nlb3.py]: s/2017/2018/

File: tools/extra/pydiag/opae/tools/fpgadiag/nlb7.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
+# Copyright(c) 2017, Intel Corporation
```
Comment (tswhison) [inline tools/extra/pydiag/opae/tools/fpgadiag/nlb7.py]: s/2017/2018/

File: tools/extra/pydiag/opae/utils/byteutils.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
+#! /usr/bin/env python
+# -*- coding: utf-8 -*-
+# vim:fenc=utf-8
+#
+# Copyright © 2018 rrojo <rrojo@ub-rojo-vm-04.amr.corp.intel.com>
```
Comment (tswhison) [inline tools/extra/pydiag/opae/utils/byteutils.py]: Remove? or replace with Intel (c) header?

File: tools/extra/pydiag/opae/utils/csr.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
+# Copyright(c) 2017, Intel Corporation
```
Comment (tswhison) [inline tools/extra/pydiag/opae/utils/csr.py]: s/2017/2018/

File: tools/extra/pydiag/opae/utils/csr.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    in the format of (hi, lo).
+    __types__ is a cache of csr derived types that have already been
+    generated"""
+    _bits_ = []
+    _offset_ = 0x0000
+    _types_ = {}
+    _enums_ = []
+
+    def __init__(self, value=0, offset=None, **kwargs):
+        """__init_ Create a new csr object and initialize it with the given
```
Comment (tswhison) [inline tools/extra/pydiag/opae/utils/csr.py]: s/__init_/__init__/

File: tools/extra/pydiag/opae/utils/csr.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        """
+        self._offset = self._offset_ if offset is None else offset
+        type_name = self.__class__.__name__
+        if type_name in self._types_:
+            union_t = self._types_[type_name]
+        else:
+            self._width = kwargs.get('width', 64)
+            # if the class has defined bitfields
+            if self._bits_:
+                # sort the bitfields by the second element, bit bit range
```
Comment (tswhison) [inline tools/extra/pydiag/opae/utils/csr.py]: s/bit bit/bit/

File: tools/extra/pydiag/opae/utils/csr.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+                self._bits_ = sorted(self._bits_, key=lambda b: b[1])
+                # the smallest bit is the last bitfield in the sorted list
+                # the bitfield is make of of the name, bitrange
```
Comment (tswhison) [inline tools/extra/pydiag/opae/utils/csr.py]: s/of of/of/

File: tools/extra/pydiag/opae/utils/utils.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
+#! /usr/bin/env python
+# Copyright(c) 2017, Intel Corporation
```
Comment (tswhison) [inline tools/extra/pydiag/opae/utils/utils.py]: s/2017/2018/

File: tools/extra/pydiag/setup.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+from setuptools import setup, find_packages
```
Comment (tswhison) [inline tools/extra/pydiag/setup.py]: Needs a (c) header.

File: tools/CMakeLists.txt [hunk 11]
```diff
@@ -36,7 +36,7 @@ add_subdirectory(extra/coreidle)
 add_subdirectory(extra/userclk)
 add_subdirectory(extra/packager)
 add_subdirectory(extra/c++utils)
-add_subdirectory(extra/fpgadiag)
+add_subdirectory(extra/pydiag)
```
Comment (anandaravuri) [inline tools/CMakeLists.txt]: change name pydiag to pyfpgadiag or fpadiagpy

File: tools/extra/pydiag/opae/tools/fpgadiag/diagtest.py [hunk 12]
```diff
@@ -0,0 +1,440 @@
...
+        stream_handler.setFormatter(logging.Formatter(
+            '%(asctime)s: [%(name)-8s][%(levelname)-6s] - %(message)s'))
+        self.logger.addHandler(stream_handler)
+
+        parser.add_argument(
+            "-c",
+            "--config",
+            help="Path to test config file")
+        parser.add_argument("-t", "--target", help="one of {fpga, ase}")
+        parser.add_argument(
```
[the remark this answers is not in the document]
Comment (r-rojo) [reply tools/extra/pydiag/opae/tools/fpgadiag/diagtest.py]: We can probably do it later. It shouldn't block this PR I think

**Comment by tswhison** [comment]
pyopae/setup.py needs a (c) header as well.

**Pull request merged**